[Refactor] 퀴즈 세트 조회 기능 리팩토링

Hut234 opened the pull request:
## 🔖 연관된 이슈
<!-- 연관된 이슈 번호를 작성해 주세요 -->
#41 

## 📂 작업 내용
<!-- 작업한 내용을 작성해 주세요 -->
- 퀴즈 세트, 퀴즈, 퀴즈 보기에 컬럼 추가
- 퀴즈 세트 조회 시 퀴즈 시작 번호를 쿼리 파라미터로 입력 받도록 컨트롤러 수정
- 변경된 내용을 반영하여 테스트 코드 수정

## 📢 리뷰 참고사항
<!-- 리뷰시 참고사항을 작성해 주세요 -->

Hut234 committed:
feat: Add columns to QuizSet, Quiz, QuizOption Entity (added column: estimatedTime, totalQuizCount, sequence)

Hut234 committed:
chore: Add Flyway migration file for adding estimated_time, total_quiz_count, sequence column to quiz_set, quiz, quiz_option table

Hut234 committed:
refactor: Support startQuizSequence query param for partial quiz response

Hut234 committed:
chore: Remove unused GetUserQuizSetResponse class

POKUDING commented on src/main/java/com/readup/server/quiz/domain/model/QuizSet.java:
```diff
@@ -52,22 +52,35 @@ public class QuizSet extends BaseEntity {
 	}
 
+	private static int calculateEstimatedTime(int totalQuizCount) {
+		return totalQuizCount * 2;
```
나중에 실제 시간으로 계산할 경우를 생각하면 int 자료 구조가 맞을지 우려됩니다!
Duration 자료형을 이용하는게 어떨지 건의 드립니다!

Hut234 replied:
디자인 기준으로 예상 퀴즈 소요 시간의 최소 단위는 분 단위로 보여서 int 타입으로 정의하였습니다.
추후 예상 퀴즈 소요 시간 로직을 고도화할 때 비즈니스 로직 or 유저가 머무른 시간을 나타낼 테이블의 컬럼에서 Duration 타입을 사용해서 계산하는 방식으로 사용하고 DB에 저장하는 타입은 int 타입으로 괜찮다고 판단하였습니다.

POKUDING commented on src/main/resources/db/migration/V1_5_4__add_columns_to_quiz_set_quiz_quiz_option.sql:
```diff
@@ -0,0 +1,9 @@
+ALTER TABLE `quiz_set`
+    ADD COLUMN `estimated_time` INT NOT NULL,
```
앞서 말씀드린 Duration 자료형을 MySQL에서는 BIGINT 혹은 INTERVAL 을 사용한다고 합니다.
저도 자세한 내용은 모르지만 알아보고 공유드리겠습니다!

POKUDING commented on src/main/java/com/readup/server/quiz/domain/model/QuizSet.java:
```diff
@@ -52,22 +52,35 @@ public class QuizSet extends BaseEntity {
 			.likeAverage(0.0)
 			.correctAnswerAverage(0.0)
+			.estimatedTime(calculateEstimatedTime(totalQuizCount))
+			.totalQuizCount(totalQuizCount)
```
totalQuizCount 의 경우 외부 주입 보다 QuizList 의 size 로 계산되면 어떨까요?

Hut234 replied:
QuizList의 size 값을 활용할 수 있도록 메서드를 분리하여 수정에 반영하였습니다.

POKUDING commented on src/main/java/com/readup/server/quiz/presentation/QuizSetController.java:
```diff
@@ -28,7 +29,7 @@ public ApiResponse<CreateQuizSetResponse> createQuizSet(@RequestBody CreateQuizS
 	@GetMapping("/private/quiz-sets/{quizSetId}")
-	public ApiResponse<GetQuizSetResponse> getQuizSet(@PathVariable Long quizSetId) {
-		return successResponse(quizSetService.getQuizSet(quizSetId));
+	public ApiResponse<GetQuizSetResponse> getQuizSet(@PathVariable Long quizSetId, @RequestParam(defaultValue = "1") int startQuizSequence) {
```
startQuizSequence 에 음수 처리가 되어도 좋을 것 같습니다.

ex)
컨트롤러 @Validated 어노테이션 추가
파라미터 @Min(1) 어노테이션 추가

Hut234 replied:
음수값에 대한 유효성 처리를 반영하였습니다.

POKUDING reviewed (approved):
고생 많으셨습니다!! 
깔끔하게 수정해 주셔서 이해하기 좋았습니다.
현재도 좋아보이나 코멘트 남겨둔 부분 한번 확인해 주시면 감사하겠습니다 ㅎㅎ

Hut234 committed:
refactor
- Extract totalQuizCount and estimatedTime update logic from quizSet creation flow into a separate method
- Extract calculation logic into static QuizSetTimeCalculator class

Hut234 committed:
feat
- Add validation for startQuizSequence query parameter
- Add HandlerMethodValidationException handler
- Add MethodArgumentNotValidException handler
